### DISCUSSION TITLE
Designer.bat is file is not generating xpath expression

### DISCUSSION THREAD

**Discussion by RaheemShaik999** [pmd/pmd#3035] [Q&A] [answered]
Hi Github,
I am facing issue with the pmd designer.bat file, it is not generating xpath for my java code.
Anyone please help me.

**Comment by adangel** [comment]
Hi @RaheemShaik999,
which version of PMD are you using?
Do you get any error message?

Please note, the designer can't generate an xpath expression for you - you'll need to write the expression yourself. It however assists you in writing it (via auto complete).
At the end, you can export the xpath expression as a rule definition via menu "File" -> "Export XPath to rule". In the dialog, you can fill out any remaining detail and then copy-and-paste the rule-xml snippet into your custom ruleset.

**Comment by RaheemShaik999** [comment]
@andreas, Thanks.
i am using pmd 6.6.
i am not getting any error message.
Now my questing is
which is better for generating custom rules out of java or xpath ?
if am using xpath, where i can add the xpath?
bit confusion all these.
if you have any sample project for pmd custom rules, pls send.
Thanks in advance.

**Comment by adangel** [comment, marked as answer]
Hi @RaheemShaik999,
you're right, there are two ways of writing custom rules: XPath based rules and Java Rules. They are described in our documentation here: https://pmd.github.io/pmd-6.7.0/pmd_userdocs_extending_writing_pmd_rules.html and https://pmd.github.io/pmd-6.7.0/pmd_userdocs_extending_writing_xpath_rules.html . 
XPath rules are usually simpler (e.g. they cannot maintain state) and Java-based rules are most flexible. XPath rules can be used easier in a custom ruleset, since the rule is entirely defined in XML. See also our doc on custom rulesets: https://pmd.github.io/pmd-6.7.0/pmd_userdocs_making_rulesets.html

A example for a custom java-based rule can be found here: https://github.com/pmd/pmd-examples

**Comment by RaheemShaik999** [comment]
I went through your example project, when i try to setup the project what you provided by using cmd.  It is showing following errors:
1. net.sourceforge.pmd.pmd error in the cmd screen
2. Abstract java rule showing error - 
The type net.sourceforge.pmd.lang.rule.AbstractRule cannot be resolved. It is indirectly referenced from required .class files	WhileLoopsMustUseBracesRule.java	/P
few things i need answers, pls provide it can helpful - 
is there any need to add existing pmd6.6 jar in my new customrules jar because we are extending abstractjavarule right ?
How to add my new rules to existing pmd6.6 jar ?
I am using eclipse, which project i have to take for creating new custom rules ?

**Comment by adangel** [comment]
Hi, yes, I've checked it - there were a couple of errors (I've just fixed them now in the readme):
* when using `~code`, this does not work - it should be `~/code`
* the git clone command was not complete
* the javac command is missing one "..", so it didn't find the folder, where pmd has been extracted. the correct command is:
`javac -d build -cp '../../pmd-bin-6.0.1/lib/*' src/*.java`
* the same applies when executing pmd later, we need to go two levels up:
`CLASSPATH=custom-rule-example.jar ../../pmd-bin-6.0.1/bin/run.sh pmd -f text -d testsrc -R myrule.xml`

Here are the whole changes: https://github.com/pmd/pmd-examples/commit/74e40b9f50e63ddb39ad0660dd555c75d26058f8#diff-8c9691456fd6146cb1ee643c976b041a

**Comment by RaheemShaik999** [comment]
Hi...I am running rules by using cmd.
My question is - 
1. At the time of creating custom rules we are extending abstractjavarule to our new rule. For that how I can add that super class. Because it is in the format of binary.....So bit confusion in it.

**Comment by adangel** [comment]
Okay, extending `AbstractJavaRule` means, you are going down the "Java-based rule" approach. See https://github.com/pmd/pmd-examples/blob/master/custom-rule-manually/src/MyRule.java as an example.

**Comment by RaheemShaik999** [comment]
Ok...but in the MyRule.java, following packages where I can get: 
import net.sourceforge.pmd.lang.java.ast.ASTVariableDeclaratorId;

import net.sourceforge.pmd.lang.java.rule.AbstractJavaRule;

**Comment by adangel** [comment]
You can also use https://github.com/pmd/pmd-examples/tree/master/custom-rule-with-maven and import this with your favorite Java IDE (eclipse, intellij, netbeans, ...)

**Comment by RaheemShaik999** [comment]
I have used the above project as u mentioned, but it is showing the following error:
could not find or load main class net.sourceforge.pmd.PMD
i have used following command:
C:\Backup\pmd-bin-6.6.0\bin\pmd.bat -d C:\Data\Workspaces\Practice\Mavendemo\test\src -f xml -R rulesets\java\myrule.xml

pmd.bat file contains the following:
@echo off
set TOPDIR=%~dp0..
set OPTS=
set MAIN_CLASS=net.sourceforge.pmd.PMD
java -classpath "%TOPDIR%\lib\*" %OPTS% %MAIN_CLASS% %*

**Comment by adangel** [comment]
Can you modify the file `pmd.bat` and remove the line with `@echo off` and try again? Then we should see the actual paths, that are used.

The error "could not find or load main class net.sourceforge.pmd.PMD" means, that the "classpath" argument is wrong. It should point to the lib-subfolder in the directory, where you have extracted the pmd zip file, so that java finds all the jar files. In your case, it should be `C:\Backup\pmd-bin-6.6.0\lib*`.

**Comment by RaheemShaik999** [comment]
Thanks for your suggestion,,.
But am getting the exception: Rulesetnotfoundexception.
pom.xml:

```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<groupId>sr</groupId>
	<artifactId>test</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<packaging>jar</packaging>
	<name>test</name>
	<url>http://maven.apache.org</url>

	<properties>
		<pmd.version>6.0.1</pmd.version>
		<java.version>1.7</java.version>
		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
	</properties>

	<dependencies>
		<dependency>
			<groupId>net.sourceforge.pmd</groupId>
			<artifactId>pmd-java</artifactId>
			<version>${pmd.version}</version>
			<scope>compile</scope>
		</dependency>
		<dependency>
			<groupId>junit</groupId>
			<artifactId>junit</artifactId>
			<version>3.8.1</version>
			<scope>test</scope>
		</dependency>
	</dependencies>
</project>
```

My folder structure is like:
C:\Data\Workspaces\Practice\Mavendemo\test\src\main\java\sr\myrules\myrule.xml
C:\Data\Workspaces\Practice\Mavendemo\test\src\main\java\sr\test\MyRule.java

**Comment by adangel** [comment]
Hi @RaheemShaik999 , do you still have this issue?

**Comment by adangel** [comment]
If you still have trouble running PMD, please open a new issue.
